### PR TITLE
MCR-3780 Create common base class for source implementations

### PR DESCRIPTION
**Pull request by toKrause**
[Link to jira](https://mycore.atlassian.net/browse/MCR-3780).

# Pull Request Checklist (Author)

Please go through the following checklist before assigning the PR for review:

## Ticket & Documentation
- [x] The issue in the ticket is clearly described and the solution is documented.
- [x] Design decisions (if any) are explained.
- [x] The ticket references the correct source and target branches.
- [x] The `fixed-version` is correctly set in the ticket and matches the PR's target branch (`main`).

## Feature & Improvement Specific Checks
- [ ] Instructions on how to test or use the feature are included or linked (e.g. to documentation).
- [ ] For UI changes: before & after screenshots are attached.
- [ ] New features or migrations are documented.
- [ ] Does this change affect existing applications, data, or configurations?
  - [ ] Yes: Is a migration required? If yes, describe it.
  - [ ] Breaking change is marked in the **commit message**.

## Testing
- [x] I have tested the changes locally.
- [x] The feature behaves as described in the ticket.
- [ ] Were existing tests modified?
  - [ ] Yes: explain the changes for reviewers.

## MCR Conventions & Metadata
- [x] [MCR naming conventions](https://www.mycore.de/documentation/developer/conventions/) are followed
- [ ] If the **public API** has changed:
  - [ ] Old API is deprecated or a migration is documented.
  - [x] If not, no action needed.
- [x] Java license headers are added where necessary.
- [x] Javadoc is written for non-self-explanatory classes/methods (Clean Code).
- [x] All configuration options are documented in Javadoc and `mycore.properties`.
- [x] No default properties are hardcoded — all set via `mycore.properties`.

## Multi-Repo Considerations
- [ ] Is an equivalent PR in `MIR` required?
  - [ ] If yes, is it already created?

### CODE REVIEW

File: mycore-base/src/main/java/org/mycore/common/config/instantiator/source/MCRSourceBase.java [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+        return instance;
+
+    }
+
+    protected final boolean rejectedBySentinel(MCRSentinel sentinel, MCRSourceContext context,
+        Map<String, String> properties, String prefix) {
+
+        if (sentinel != null) {
+            boolean sentinelValue = sentinel.defaultValue();
+            String configuredSentinelValue = properties.remove(prefix + sentinel.name());
```
Comment (Mewel) [inline mycore-base/src/main/java/org/mycore/common/config/instantiator/source/MCRSourceBase.java]: Muss hier wirklich der Wert entfernt werden oder reicht ein einfaches get()? Was ist wenn du eine Unmodifiable Map hier rein bekommst, dann würde das knallen mit UnsupportedOperationException.

**Review by Mewel — CHANGES_REQUESTED** [review]
Kannst du diesen Test noch mit einbauen (Claude & Codex):

```
/**
 * Tests the interaction of {@link MCRSentinel} with the different lookups performed for a {@link MCRProperty},
 * for the following conditions:
 * <ol>
 *   <li>The property is looked up
 *   <ol>
 *     <li>relative to the name of the configured instance,</li>
 *     <li>as an absolute property (i.e. <code>absolute = true</code>),</li>
 *     <li>as a default property (i.e. <code>defaultName = "..."</code>, with the relative property not set)</li>
 *   </ol>
 *   <li>Annotation has
 *   <ol>
 *     <li>no sentinel,</li>
 *     <li>an enabling sentinel, but no configured sentinel value (implicitly enabling sentinel)</li>
 *     <li>a sentinel and an enabling configured sentinel value (explicitly enabling sentinel)</li>
 *     <li>a disabling sentinel, but no configured sentinel value (implicitly disabling sentinel)</li>
 *     <li>a sentinel and a disabling configured sentinel value (explicitly disabling sentinel)</li>
 *   </ol>
 * </ol>
 * <table style="border-collapse: collapse;">
 *   <caption>Expected results for different conditions (all properties are optional)</caption>
 *   <tr>
 *     <th style="border: 1px solid;">Lookup</th>
 *     <th style="border: 1px solid;">Sentinel</th>
 *     <th style="border: 1px solid;">Expected Result</th>
 *   </tr>
 *   <tr>
 *     <td style="border: 1px solid;">relative or absolute</td>
 *     <td style="border: 1px solid;">none or enabled</td>
 *     <td style="border: 1px solid;"><code>Value</code></td>
 *   </tr>
 *   <tr>
 *     <td style="border: 1px solid;">relative or absolute</td>
 *     <td style="border: 1px solid;">impl. or expl. disabled</td>
 *     <td style="border: 1px solid;"><code>null</code></td>
 *   </tr>
 *   <tr>
 *     <td style="border: 1px solid;">default</td>
 *     <td style="border: 1px solid;">none or enabled</td>
 *     <td style="border: 1px solid;"><code>Value</code></td>
 *   </tr>
 *   <tr>
 *     <td style="border: 1px solid;">default</td>
 *     <td style="border: 1px solid;">impl. or expl. disabled</td>
 *     <td style="border: 1px solid;">Exception</td>
 *   </tr>
 * </table>
 */
@MyCoReTest
public class MCRInstantiatorPropertySentinelTest {

    private static final Logger LOGGER = LogManager.getLogger();

    public static final String CONFIGURED_CLASS_PROPERTY = "Foo.Class";

    public static final String RELATIVE_VALUE_PROPERTY = "Foo.Value";

    public static final String RELATIVE_SENTINEL_PROPERTY = "Foo.Value.Enabled";

    public static final String ABSOLUTE_VALUE_PROPERTY = "MCR.Absolute.Value";

    public static final String ABSOLUTE_SENTINEL_PROPERTY = "MCR.Absolute.Value.Enabled";

    public static final String DEFAULT_VALUE_PROPERTY = "MCR.Default.Value";

    public static final String DEFAULT_SENTINEL_PROPERTY = "MCR.Default.Value.Enabled";

    private static Stream<Arguments> provideAllParameterCombinations() {
        List<Arguments> argumentsList = new ArrayList<>();
        for (Lookup lookup : Lookup.values()) {
            for (Sentinel sentinel : Sentinel.values()) {
                argumentsList.add(Arguments.of(lookup, sentinel));
            }
        }
        return argumentsList.stream();
    }

    @ParameterizedTest
    @MethodSource("provideAllParameterCombinations")
    @MCRTestConfiguration(properties = {
        @MCRTestProperty(key = RELATIVE_VALUE_PROPERTY, empty = true),
        @MCRTestProperty(key = ABSOLUTE_VALUE_PROPERTY, empty = true),
        @MCRTestProperty(key = DEFAULT_VALUE_PROPERTY, empty = true)
    })
    void test(Lookup lookup, Sentinel sentinel) {

        // log all parameters
        LOGGER.info("TEST PARAMETERS");
        LOGGER.info("lookup={}", lookup);
        LOGGER.info("sentinel={}", sentinel);

        // select class to be configured
        Class<? extends Configurable> configuredClass = switch (lookup) {
            case RELATIVE -> switch (sentinel) {
                case NONE -> RelativeNoSentinel.class;
                case IMPLICITLY_ENABLED, EXPLICITLY_DISABLED -> RelativeEnablingSentinel.class;
                case IMPLICITLY_DISABLED, EXPLICITLY_ENABLED -> RelativeDisablingSentinel.class;
            };
            case ABSOLUTE -> switch (sentinel) {
                case NONE -> AbsoluteNoSentinel.class;
                case IMPLICITLY_ENABLED, EXPLICITLY_DISABLED -> AbsoluteEnablingSentinel.class;
                case IMPLICITLY_DISABLED, EXPLICITLY_ENABLED -> AbsoluteDisablingSentinel.class;
            };
            case DEFAULT -> switch (sentinel) {
                case NONE -> DefaultNoSentinel.class;
                case IMPLICITLY_ENABLED, EXPLICITLY_DISABLED -> DefaultEnablingSentinel.class;
                case IMPLICITLY_DISABLED, EXPLICITLY_ENABLED -> DefaultDisablingSentinel.class;
            };
        };

        // set class property for configured instance
        MCRConfiguration2.set(CONFIGURED_CLASS_PROPERTY, configuredClass.getName());

        // set the value property for the lookup under test, unset the value properties for the other lookups
        // empty strings are default via @MCRTestProperty above, need to overwrite
        MCRConfiguration2.set(RELATIVE_VALUE_PROPERTY, lookup == Lookup.RELATIVE ? "Value" : null);
        MCRConfiguration2.set(ABSOLUTE_VALUE_PROPERTY, lookup == Lookup.ABSOLUTE ? "Value" : null);
        MCRConfiguration2.set(DEFAULT_VALUE_PROPERTY, lookup == Lookup.DEFAULT ? "Value" : null);

        // set sentinel property for the lookup under test
        String sentinelProperty = switch (lookup) {
            case RELATIVE -> RELATIVE_SENTINEL_PROPERTY;
            case ABSOLUTE -> ABSOLUTE_SENTINEL_PROPERTY;
            case DEFAULT -> DEFAULT_SENTINEL_PROPERTY;
        };
        switch (sentinel) {
            case EXPLICITLY_ENABLED -> MCRConfiguration2.set(sentinelProperty, "true");
            case EXPLICITLY_DISABLED -> MCRConfiguration2.set(sentinelProperty, "false");
            default -> {
                // no configured sentinel value
            }
        }

        // log all relevant configuration entries
        LOGGER.info("CONFIGURATION PROPERTIES");
        Map<String, String> propertiesMap = MCRConfiguration2.getPropertiesMap();
        LOGGER.info("{}={}", CONFIGURED_CLASS_PROPERTY, get(propertiesMap, CONFIGURED_CLASS_PROPERTY));
        LOGGER.info("{}={}", RELATIVE_VALUE_PROPERTY, get(propertiesMap, RELATIVE_VALUE_PROPERTY));
        LOGGER.info("{}={}", ABSOLUTE_VALUE_PROPERTY, get(propertiesMap, ABSOLUTE_VALUE_PROPERTY));
        LOGGER.info("{}={}", DEFAULT_VALUE_PROPERTY, get(propertiesMap, DEFAULT_VALUE_PROPERTY));
        LOGGER.info("{}={}", sentinelProperty, get(propertiesMap, sentinelProperty));

        // perform instantiation of configured class
        Configurable instance = null;
        MCRConfigurationException exception = null;
        try {
            instance = MCRInstanceConfiguration.ofName(Configurable.class, "Foo").instantiate();
        } catch (MCRConfigurationException e) {
            exception = e;
        }

        // all the indications a property value should not be obtained (or obtaining it should be suppressed)
        boolean shouldNotObtainValue = false;
        shouldNotObtainValue |= sentinel == Sentinel.IMPLICITLY_DISABLED;
        shouldNotObtainValue |= sentinel == Sentinel.EXPLICITLY_DISABLED;

        if (lookup == Lookup.DEFAULT && shouldNotObtainValue) {

            // a default property that is suppressed by a sentinel is treated as a missing default property
            assertNull(instance);
            assertNotNull(exception);

            assertEquals("Default property, configured in " + DEFAULT_VALUE_PROPERTY + " (and its sub-properties),"
                + " for target field 'value' in configured class " + configuredClass.getName()
                + " is missing", exception.getMessage());

        } else {

            assertNull(exception);
            assertNotNull(instance);

            if (shouldNotObtainValue) {
                assertNull(instance.value());
            } else {
                assertEquals("Value", instance.value());
            }

        }

    }

    private static String get(Map<String, String> map, String key) {
        String value = map.get(key);
        return value == null ? "null" : "'" + value + "'";
    }

    enum Lookup {

        RELATIVE,

        ABSOLUTE,

        DEFAULT

    }

    enum Sentinel {

        NONE,

        IMPLICITLY_ENABLED,

        EXPLICITLY_ENABLED,

        IMPLICITLY_DISABLED,

        EXPLICITLY_DISABLED

    }

    public interface Configurable {

        String value();

    }

    public static class RelativeNoSentinel implements Configurable {

        @MCRProperty(name = "Value", required = false)
        public String value;

        @Override
        public String value() {
            return value;
        }

    }

    public static class RelativeEnablingSentinel implements Configurable {

        @MCRSentinel
        @MCRProperty(name = "Value", required = false)
        public String value;

        @Override
        public String value() {
            return value;
        }

    }

    public static class RelativeDisablingSentinel implements Configurable {

        @MCRSentinel(defaultValue = false)
        @MCRProperty(name = "Value", required = false)
        public String value;

        @Override
        public String value() {
            return value;
        }

    }

    public static class AbsoluteNoSentinel implements Configurable {

        @MCRProperty(name = ABSOLUTE_VALUE_PROPERTY, absolute = true, required = false)
        public String value;

        @Override
        public String value() {
            return value;
        }

    }

    public static class AbsoluteEnablingSentinel implements Configurable {

        @MCRSentinel
        @MCRProperty(name = ABSOLUTE_VALUE_PROPERTY, absolute = true, required = false)
        public String value;

        @Override
        public String value() {
            return value;
        }

    }

    public static class AbsoluteDisablingSentinel implements Configurable {

        @MCRSentinel(defaultValue = false)
        @MCRProperty(name = ABSOLUTE_VALUE_PROPERTY, absolute = true, required = false)
        public String value;

        @Override
        public String value() {
            return value;
        }

    }

    public static class DefaultNoSentinel implements Configurable {

        @MCRProperty(name = "Value", defaultName = DEFAULT_VALUE_PROPERTY, required = false)
        public String value;

        @Override
        public String value() {
            return value;
        }

    }

    public static class DefaultEnablingSentinel implements Configurable {

        @MCRSentinel
        @MCRProperty(name = "Value", defaultName = DEFAULT_VALUE_PROPERTY, required = false)
        public String value;

        @Override
        public String value() {
            return value;
        }

    }

    public static class DefaultDisablingSentinel implements Configurable {

        @MCRSentinel(defaultValue = false)
        @MCRProperty(name = "Value", defaultName = DEFAULT_VALUE_PROPERTY, required = false)
        public String value;

        @Override
        public String value() {
            return value;
        }

    }

}

```

Den gleichen Test sollte es auch noch für MCRPropertyList und MCRPropertyMap geben.

**Comment by toKrause** [comment]
@Mewel: I've added "absolute vs. relative" lookup as an additional parameter to the existing test. The interaction with default properties is already being tested.

This should make it easier to maintain the test classes and also emphasizes how little the differences between absolute and relative lookup are.

**Comment by toKrause** [comment]
I'll look into integrating different sentinel configurations as well.

**Comment by toKrause** [comment]
@Mewel: I've switched from `remove` to `get` as requested.

With regards to the tests: There are currently some edge cases with some sentinel configurations and explicitly empty default values that make writing tests for property maps and lists extremely tricky.

All of those edge cases will vanish after [MCR-3784](https://github.com/MyCoRe-Org/mycore/pull/3066) and [MCR-3781](https://github.com/MyCoRe-Org/mycore/pull/3063) have been merged. So, spending additional time on this is a waste of time for me.

I'd like to merge this PR as is (with tests considering absolute properties) and add/extend tests for various sentinel configurations, once those two PRs have been merged.

[MCR-3784]: https://mycore.atlassian.net/browse/MCR-3784?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[MCR-3781]: https://mycore.atlassian.net/browse/MCR-3781?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

**Comment by Mewel** [comment]
I don't see the change yet. It's still .remove and not .get. Or do i see it wrong?

**Comment by toKrause** [comment]
maybe, just maybe, I forgot to push :)

**Comment by toKrause** [comment]
now everything should be okay

**Review by Mewel — APPROVED** [review]
